lib/apiservers/engine/backends/container.go

```diff
@@ -1589,13 +1589,13 @@ func (c *Container) containerAttach(name string, ca *backend.ContainerAttachConf
 
 	actor := CreateContainerEventActorWithAttributes(vc, map[string]string{})
 	EventService().Log(containerAttachEvent, eventtypes.ContainerEventType, actor)
-	defer func() {
-		actor := CreateContainerEventActorWithAttributes(vc, map[string]string{})
-		EventService().Log(containerDetachEvent, eventtypes.ContainerEventType, actor)
-	}()
 	err = c.containerProxy.AttachStreams(context.Background(), vc, clStdin, clStdout, clStderr, ca)
 	if err != nil {
 		if _, ok := err.(DetachError); ok {
+			defer func() {
```

Comment:
I think it's actually viable to dispatch the event directly rather than as a `defer` - the remaining work has to occur to tidy up, but the detach has already happened from the client's point of view, as the AttachStreams have been closed.